### PR TITLE
Geo/12445 precalculate integration coefficients

### LINKED ISSUE #12445: [GeoMechanicsApplication] Calculate integration coefficients prior to integration point loop
As a developer, I would like to calculate the integration coefficients prior to carrying out the integration point loop.

**Background**
There are 2 instances for diff order and 7 instances for non-diff-order where we need to replace the call in the integration point loop with pre-calculated lists.

**Acceptance Criteria**
- There are no places in the elements where integration coefficients are calculated for a single integration point (except for the calculation of the list for all integration points)

### PR DESCRIPTION
**Pull request by indigocoral**
**📝 Description**
Calculation of the integration coefficients prior to carrying out the integration point loop

**🆕 Changelog**
- Added pre-calculation of the integration coefficients to carry out the integration point loop for the following files:
 
_1. small_strain_U_Pw_diff_order_element.cpp
2. U_Pw_small_strain_element.cpp
3. U_Pw_small_strain_FIC_element.cpp
4. U_Pw_updated_lagrangian_element.cpp
5. U_Pw_updated_lagrangian_FIC_element.cpp
6. U_Pw_small_strain_element.cpp
7. U_Pw_small_strain_interface_element.cpp
8. updated_lagrangian_U_Pw_diff_order_element.cpp_

- Removed the CalculateIntegrationCoefficient method and inlined its functionality directly into the CalculateIntegrationCoefficients method in the _U_Pw_base_element.cpp_ and _small_strain_U_Pw_diff_order_element.cpp_ files.
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

### CODE REVIEW

File: applications/GeoMechanicsApplication/custom_elements/U_Pw_base_element.cpp [hunk 1]
```diff
@@ -491,22 +491,14 @@ void UPwBaseElement<TDim, TNumNodes>::CalculateAll(MatrixType&        rLeftHandS
...
 
 //----------------------------------------------------------------------------------------
-template <unsigned int TDim, unsigned int TNumNodes>
-double UPwBaseElement<TDim, TNumNodes>::CalculateIntegrationCoefficient(const GeometryType::IntegrationPointType& rIntegrationPoint,
-                                                                        double detJ) const
-
-{
-    return mpStressStatePolicy->CalculateIntegrationCoefficient(rIntegrationPoint, detJ, GetGeometry());
-}
-
```
Comment (avdg81) [inline applications/GeoMechanicsApplication/custom_elements/U_Pw_base_element.cpp]: This cleans up nicely. Please, also remove the corresponding declaration from the header file. In other words remove
```
double CalculateIntegrationCoefficient(const GeometryType::IntegrationPointType& rIntegrationPoint,
                                           double detJ) const;
```
from file `U_Pw_base_element.hpp`.

File: applications/GeoMechanicsApplication/custom_elements/U_Pw_small_strain_interface_element.cpp [hunk 2]
```diff
@@ -1279,8 +1282,7 @@ void UPwSmallStrainInterfaceElement<TDim, TNumNodes>::CalculateMaterialStiffness
         mConstitutiveLawVector[GPoint]->CalculateMaterialResponseCauchy(ConstitutiveParameters);
 
         // Compute weighting coefficient for integration
```
Comment (avdg81) [inline applications/GeoMechanicsApplication/custom_elements/U_Pw_small_strain_interface_element.cpp]: This comment is no longer accurate, so I would remove it.

File: applications/GeoMechanicsApplication/custom_elements/U_Pw_small_strain_element.cpp [hunk 3]
```diff
@@ -214,15 +214,16 @@ void UPwSmallStrainElement<TDim, TNumNodes>::CalculateHydraulicDischarge(const P
...
                                                    Variables.detJContainer, mThisIntegrationMethod);
+    const auto integration_coefficients =
+        this->CalculateIntegrationCoefficients(IntegrationPoints, Variables.detJContainer);
 
     // Loop over integration points
     for (unsigned int GPoint = 0; GPoint < NumGPoints; ++GPoint) {
         noalias(Variables.GradNpT) = Variables.DN_DXContainer[GPoint];
         Variables.detJ             = Variables.detJContainer[GPoint];
 
         // Compute weighting coefficient for integration
```
Comment (avdg81) [inline applications/GeoMechanicsApplication/custom_elements/U_Pw_small_strain_element.cpp]: This comment is no longer accurate, so I would remove it.

**Review by markelov208 — DISMISSED** [review]
Hi Marjan, I have more time than Anne and Richard today. From my view-point, everything is very well done. There is a duplication of CalculateIntegrationCoefficients but this is a topic of another issue.

File: applications/GeoMechanicsApplication/custom_elements/small_strain_U_Pw_diff_order_element.cpp [hunk 4]
```diff
@@ -1489,19 +1496,13 @@ std::vector<Matrix> SmallStrainUPwDiffOrderElement::CalculateBMatrices(
     return result;
 }
 
-double SmallStrainUPwDiffOrderElement::CalculateIntegrationCoefficient(const GeometryType::IntegrationPointType& rIntegrationPoint,
-                                                                       double detJ) const
-{
-    return mpStressStatePolicy->CalculateIntegrationCoefficient(rIntegrationPoint, detJ, GetGeometry());
-}
```
Comment (avdg81) [inline applications/GeoMechanicsApplication/custom_elements/small_strain_U_Pw_diff_order_element.cpp:1501]: Similar to the work you have done for class `UPwBaseElement`, this cleans up nicely. But also here I think you have forgotten to remove the declaration from the header file (i.e. `small_strain_U_Pw_diff_order_element.hpp`). Could you please take care of that?

**Review by avdg81 — COMMENTED** [review]
Thank you so much for the cleanup you've carried out. I'm satisfied with the changes you've made. As you can see for yourself, I only have a few minor suggestions.

File: applications/GeoMechanicsApplication/custom_elements/U_Pw_base_element.cpp [hunk 5]
```diff
@@ -435,20 +435,15 @@ void UPwBaseElement::CalculateAll(MatrixType&        rLeftHandSideMatrix,
...
 
-double UPwBaseElement::CalculateIntegrationCoefficient(const GeometryType::IntegrationPointType& rIntegrationPoint,
-                                                       double detJ) const
-
-{
-    return mpStressStatePolicy->CalculateIntegrationCoefficient(rIntegrationPoint, detJ, GetGeometry());
-}
-
-std::vector<double> UPwBaseElement::CalculateIntegrationCoefficients(
+//----------------------------------------------------------------------------------------
```
Comment (rfaasse) [inline applications/GeoMechanicsApplication/custom_elements/U_Pw_base_element.cpp]: Could you remove this separator line? It's not really necessary and clutters the file a bit (I know we have it in other places, but we're planning to remove them)

**Review by rfaasse — COMMENTED** [review]
Nice, this cleans up the last of the pre-calculated lists that we still wanted to do, thanks for taking care of it! I have just one comment on a separator line that can be removed, next to Annes comments, nothing else!

File: applications/GeoMechanicsApplication/custom_elements/U_Pw_base_element.cpp (hunk 5 above)
Comment (indigocoral) [reply applications/GeoMechanicsApplication/custom_elements/U_Pw_base_element.cpp]: done

File: applications/GeoMechanicsApplication/custom_elements/small_strain_U_Pw_diff_order_element.cpp (hunk 4 above)
Comment (indigocoral) [reply applications/GeoMechanicsApplication/custom_elements/small_strain_U_Pw_diff_order_element.cpp:1501]: done

File: applications/GeoMechanicsApplication/custom_elements/U_Pw_small_strain_element.cpp (hunk 3 above)
Comment (indigocoral) [reply applications/GeoMechanicsApplication/custom_elements/U_Pw_small_strain_element.cpp]: done

File: applications/GeoMechanicsApplication/custom_elements/U_Pw_small_strain_interface_element.cpp (hunk 2 above)
Comment (indigocoral) [reply applications/GeoMechanicsApplication/custom_elements/U_Pw_small_strain_interface_element.cpp]: done

File: applications/GeoMechanicsApplication/custom_elements/U_Pw_base_element.cpp (hunk 1 above)
Comment (indigocoral) [reply applications/GeoMechanicsApplication/custom_elements/U_Pw_base_element.cpp]: done

**Review by avdg81 — APPROVED** [review]
Thank you very much for processing the review suggestions. I'm fully satisfied, so please feel free to merge whenever you're ready for it.

**Pull request merged**